keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,366 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Proposal to add a new kubelet gRPC API with endpoint returning local pods information.
+Serving pods statuses by kubelet within a Node will increase reliability and reduce load to the Kubernetes API Server and traffic outside the node. A connectivity issue between Node and Control Plane should not impact workloads which depend on pods readiness statuses.
+
+## Motivation
+
+Kubelet is responsible for running Health Checks (probes) and communicating the results via Pod status. All that information is stored in cache and reported to Kube-API. Right now pod information is tightly coupled with the Kubernetes API Server. When a workload wants to know the actual state of pods running on the Node, it needs to fetch it from Kube-API.
+This causes some issues:
+* Reliability - for various reasons Kube-API might not be available (connectivity issue, control plane updates)
+* Scalability - adding new watchers to kube-API is a scalability concern
+By exposing the endpoint that will serve the pods status directly from the kubelet cache we can use  it on node workloads and avoid additional dependencies on the Kubernetes API Server.
+
+| Impact | Description|
+| ------- | ------------ |
+| + | Reliability - for various reasons kube-API might not be available but this doesn’t mean that local workloads are not accessible and on node system workloads should have the most recent data about pod's readiness even when kube-API is unreachable. |
+| + | Scalability - Reduce the load on kube-API be reducing the number of watchers and using kubelet to fetch local pods data. Fetching only pods limited to one node is costly operation for kube-API. |
+| + | Safety - Read-only API will not add security risks. |
+| + | Reduce resource consumption by workload. Using Kube-API we can fetch objects like PodSpec or PodStatus, for some on-node workload this might be unnecessary, with this API workload can reduce the resource consumption. |
+| - | This API will add load to kubelet (mitigation: API will be rate limited) |
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is design to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
```

Comment:
We have discussed that at sig-arch previously, which had a fairly good representation from sig-auth. The idea is that what we're doing here is not particularly novel. We already have a Pod resources API in Kubelet that is exposed in a similar way (via socket and therefore only accessible to other workloads on the same Node). We believe that Pod readiness is safe to make available to workloads running on the same node. I think both @deads2k and @liggitt have some context here, maybe they can chime in here. 

In the future the there may be a proposal to increase the scope of the information exposed by this endpoint, but that should go through a separate audit before it is approved.

Reply:
If this is about readiness only that is acceptable, the problem I struggled with, was that in multiple places various pieces of information were alluding to more broadly exposed information:
1. in summary: `Proposal to add a new Kubelet gRPC API with endpoint returning local Pods information.`
2. in motivation section, there are multiple statements alluding to `local Pods data`
3. in goals section: `create a new API that will expose pod conditions that are computed` 
4. in risks and mitigations: `We will expose only a small portion of the pod's information related to the pod readiness.` 
5. also in risks and mitigations: `his API should serve only complete information about workloads. If Kubelet is in the init phase and not all pod’s information is known, then the API should report that information.` 
6. in control plane availability issue: `The new API should return Pods statuses that are computed by Kubelet even if
they are not reported back to kube-API but reflect the current Pod’s state locally.`
7. the API proposal calls method `ListPodConditions`, `WatchPodConditions`, `GetPodConditions`

I'd suggest rewording those places, to ensure that it's explicitly stated the scope of the expose information. My current confusion came from the fact that various places mentioned various scopes of exposed data, once it was called conditions, other places - pod status, etc. Where each of that information seems to me more than just pod readiness you're talking about.

Reply:
Good point. Let's just make sure that we explicitly mention that future APIs expansion is expected. This API is limited to pod readiness in this KEP, but open for future KEPs. For example, we wouldn't name the socket `pod-readiness` and will use way more generic name.

Reply:
@soltysh @SergeyKanzhelev I updated the doc and changed all those places to point out that we want to expose readiness information. Regarding the methods name I will leave it as Conditions since at this phase the Pod's condition form v1.PodStatus will be exposed via this API. 
With this API we don't want to just expose the boolean value of Ready status (True or False) but more granular state of the pod's readiness which is reflected in those Conditions.

Reply:
For API methods it's fine to have them built with expansion in mind, but as long as the rest of the document, and the comments on those method are explicit about what functionality is available it should be all good.

Reply:
+1 and will take another pass.